test/test_transformers.py

```diff
@@ -3030,7 +3030,7 @@ def test_flash_attention_vs_math_ref_grads(self, device, batch_size: int, seq_le
             *zip(grads_ref, grads_ref_lp, grads),
             fudge_factors={
                 'out': 1.5,
-                'grad_query': 26.0,
+                'grad_query': 68.0,
```

Comment:
its a little strange to me that if the new math implementation is supposed to more closely match the flash attention kernel that it is so much more precise for grad_q on to fp64 math. 

I actually recently updated this test in prep for your PR expecting that we would not need to tolerance bump for the flash impl especially by so much

Reply:
Check comment in https://www.internalfb.com/diff/D58710805?dst_version_fbid=437617645906478&transaction_fbid=1233726084662163 . It's a bit frustrating to make a simple change for SDPA with almost 30 times rebase :)

`grads_ref_lp` returned from Low Precision Math Reference has been significantly improved with this PR. Before this PR, let's say it's grad_q_ref_rtol=3142 , and after this PR, it's 0.08.

With this comparison logic, originally before this PR we have the boundary of 26.0 * 3142 (grad_q_ref_rtol)  . Now after this PR the boundary is 26.0 * 0.08 (grad_q_ref_rtol). Even if I increase the ratio from 26.0 to 100.0, the absolute error bound will be much less than the original error bound before this PR.

Reply:
Yeah its a very popular piece of code ;)

I agree with the original premise that the 'rtol' was too lenient thats why I switched the testing to instead only measure against max abs difference. But the updated tests now:

1.) Uses fp64 as the gold standard and computes using the math impl.
2.) Gets a threshold by calculating the max abs difference between `low precision` math and fp64 math.
3.) Asserts the fused impls outs and grads' max abs difference between it and the gold impl is less than the tolerance from 2.

In the current impl ( before this PR). There are intermediate casts to the `low precision` dtype for the `low precision` math impl on fp16 and bf16 inputs. If the fused impls were uniformly more precise than the current math impl (since they dont do these roundings) then I would expect that the fudge factors would all be `1.0`. That however was not the case. So when you bump the tolerances now you are only affecting the absolute error not the relative.

There is a great note connecting the errors introduced in FAv2 to the iterative softmax algorithm: https://fb.workplace.com/notes/200492239686121

TBH, these unit tests are not designed to 'prove' the accuracy of the fused kernels, that is proved out in training and in inference in E2E use cases, since so far we havent been taking an ULP like framing of these ops(maybe we should). 

The main goal is to catch any regressions that might get added and to be sure to cover a large swash of the fused kernels configuration space. If these changes still ensure that I am okay. If you are saying that we made the math's grad query 30x closer to the fp64 version and thus we need to bump flash's grad query to be 30x the new tolerance I am okay with that.